cmd/clusterctl/pkg/log/logger.go

```diff
@@ -0,0 +1,281 @@
+	return NewLoggerWithChannel(nil, threshold)
+}
+
+// NewLogger returns a new instance of the clusterctl logger ready for sending log entries to a channel.
+func NewLoggerWithChannel(ch LogChannel, threshold int) logr.Logger {
+	return &Logger{
+		threshold: threshold,
+		ch:        ch,
+		level:     0,
+		prefix:    "",
+		values:    nil,
+	}
+}
+
+// Logger defines a clusterctl friendly logr.Logger
+type Logger struct {
+	threshold int
+	ch        LogChannel
+	level     int
+	prefix    string
+	values    []interface{}
+}
+
+var _ logr.Logger = &Logger{}
+
+// Enabled tests whether this Logger is enabled.
+func (l *Logger) Enabled() bool {
+	return l.level >= l.threshold
+}
+
+// Info logs a non-error message with the given key/value pairs as context.
+func (l *Logger) Info(msg string, kvs ...interface{}) {
+	if l.Enabled() {
+		values := copySlice(l.values)
+		values = append(values, kvs...)
+		values = append(values, "msg", msg)
+		l.write(values)
+	}
+}
+
+// Error logs an error message with the given key/value pairs as context.
+func (l *Logger) Error(err error, msg string, kvs ...interface{}) {
+	values := copySlice(l.values)
+	values = append(values, kvs...)
+	values = append(values, "msg", msg, "error", err)
+	l.write(values)
+}
+
+// V returns an InfoLogger value for a specific verbosity level.
+func (l *Logger) V(level int) logr.InfoLogger {
+	new := l.clone()
+	new.level = level
+	return new
+}
+
+// WithName adds a new element to the logger's name.
+func (l *Logger) WithName(name string) logr.Logger {
+	new := l.clone()
+	if len(l.prefix) > 0 {
+		new.prefix = l.prefix + "/"
+	}
+	new.prefix += name
+	return new
+}
+
+// WithValues adds some key-value pairs of context to a logger.
+func (l *Logger) WithValues(kvList ...interface{}) logr.Logger {
+	new := l.clone()
+	new.values = append(new.values, kvList...)
+	return new
+}
+
+func (l *Logger) write(values []interface{}) {
+	entry := LogEntry{
+		Prefix: l.prefix,
+		Level:  l.level,
+		Values: values,
+	}
+	if l.ch != nil {
+		l.ch <- entry
+		return
+	}
+	f, err := Flatten(entry)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(f)
+}
+
+func (l *Logger) clone() *Logger {
+	return &Logger{
+		ch:     l.ch,
+		level:  l.level,
+		prefix: l.prefix,
+		values: copySlice(l.values),
+	}
+}
+
+func copySlice(in []interface{}) []interface{} {
+	out := make([]interface{}, len(in))
+	copy(out, in)
+	return out
+}
+
+// Flatten returns a human readable/machine parsable text representing the LogEntry.
+// Most notable difference with the klog implementation are:
+// - The message is printed at the beginning of the line, without the Msg= variable name e.g.
+//   "Msg"="This is a message" --> This is a message
+// - Variables name are not quoted, eg.
+//   This is a message "Var1"="value" --> This is a message Var1="value"
+// - Variables are not sorted, thus allowing full control to the developer on the output.
+func Flatten(entry LogEntry) (string, error) {
```

Comment:
`Flatten` is exported here. Are we expecting users to use this method of this package? I thought the we were aiming for `logr.Logger` interface.

Reply:
Since we are exposing LogEntry, I find useful to expose Flatten too

Reply:
If this is meant to be part of clusterctl library, we are suggesting to the user that they can build their own `LogEntry` object and use `Flatten` to print out logs.
```
le := LogEntry{
  Prefix: prefix,
  Level:  level,
  Values: values,
}
str, _  := log.Flatten(le)
fmt.Println(str)
```
If this usage is fine, then I'm ok with exposing this function.
I'm asking because I don't know how we'd like the users to use this package.